### ISSUE TITLE
bluealsa stops when bluealsa-play is launched (A2DP Sink)

### ISSUE THREAD

**Issue by mwetterw** [arkq/bluez-alsa#741]
### Problem

bluealsa quits immediately as soon as bluealsa-play is launched.

### Reproduction steps

1. Launch `btattach -P bcm -B /dev/ttyAMA0`
2. Launch `bluetoothd -n -p a2dp`
3. Launch `bluealsa -p a2dp-sink`
4. Pair, Connect, Trust with the Phone to the Raspberry Pi
5. Launch `bluealsa-aplay -v -Dplughw:CARD=Headphones --profile-a2dp`

bluealsa stops unexpectedly.

### Setup

Here, the goal is to create a Bluetooth speaker by using a Raspberry Pi.
The Raspberry Pi will be connected through the 3.5 Jack to an amplifier.
The normal usecase will be for Phones/Tablets to connect to the Raspberry Pi and stream audio there.

:warning: Note that I'm not sure at all there is a bug. My issue may come from my very specific setup.
I may miss a kernel option, a compilation flag or a configuration file.

- Completely custom embedded Linux distribution (not even Buildroot: built with my own custom tool from sources).
- Run on the Raspberry 3B (64bit build)
- With the Raspberry Pi full-featured "normal" GPU firmware (start.elf and fixup.elf and not the cutdown version)
- With the **upstream** Linux kernel compiled from the sources (and not the Raspberry Pi fork of it)
- With the BCM43438A1 blob firmware loaded into the BT chip
- My kernel build does not activate the kernel modules facility. When a feature of the kernel is enabled, it is built-in into the kernel.
- On my custom distribution, there is no dynamic loader at all, so every single lib and binary is **statically built**.
- I didn't install the bluealsa pcm and ctl libraries in alsa-lib folder. (Am I correct that it's not necessary for a simple A2DP sink scenario?)
- There is only a single user: **root** and everything runs as this user.

#### Versions

Cross-compiler **aarch64-linux-musl-gcc** 11.4.0
Libc **musl** 1.2.5

linux 6.6.63 (aarch64)

pcre2 10.44
libffi 3.4.6
sbc 2.0
fdk-aac 2.0.3
alsa-lib 1.2.13
alsa-utils 1.2.13
ncurses 6.4
readline 8.2
expat 2.6.4
dbus 1.14.10
glib 2.82.2
bluez 5.79
zlib 1.3.1
bluez-alsa 4.3.1

### Additional context

```
# aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: Loopback [Loopback], device 0: Loopback PCM [Loopback PCM]
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7
card 0: Loopback [Loopback], device 1: Loopback PCM [Loopback PCM]
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7
card 1: Headphones [bcm2835 Headphones], device 0: bcm2835 Headphones [bcm2835 Headphones]
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7
```

```
# aplay -L
null
    Discard all samples (playback) or generate zero samples (capture)
default:CARD=Loopback
    Loopback, Loopback PCM
    Default Audio Device
sysdefault:CARD=Loopback
    Loopback, Loopback PCM
    Default Audio Device
front:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    Front output / input
surround21:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    6.0 Surround output to Front and Rear speakers
surround41:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    5.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    7.0 Surround output to Front, Center and Rear speakers
surround51:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    6.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    8.1 Surround output to Front, Center, Side, Rear and Woofer speakers
default:CARD=Headphones
    bcm2835 Headphones, bcm2835 Headphones
    Default Audio Device
sysdefault:CARD=Headphones
    bcm2835 Headphones, bcm2835 Headphones
    Default Audio Device
```

```
bluealsa-aplay -l
**** List of PLAYBACK Bluetooth Devices ****
**** List of CAPTURE Bluetooth Devices ****
hci0: 60:00:01:02:03:04 [Phone], phone
  A2DP (AAC): S16_LE 2 channels 44100 Hz
```

```
bluealsa-aplay -L
bluealsa:DEV=60:00:01:02:03:04,PROFILE=a2dp,SRV=org.bluealsa
    Phone, phone, capture
    A2DP (AAC): S16_LE 2 channels 44100 Hz
````

```
# bluealsa -p a2dp-sink
bluealsa: [178] D: a2dp-aac.c:800: FDK-AAC decoder capabilities: aac=0x1a4ffff sbr=0x1bf dmx=0x7f
bluealsa: [178] D: storage.c:90: Initializing persistent storage: //lib/bluealsa
bluealsa: [178] D: main.c:663: Starting main dispatching loop
bluealsa: [178] D: main.c:118: Acquired D-Bus service name: org.bluealsa
bluealsa: [178] D: bluealsa-dbus.c:358: Registering BlueALSA D-Bus manager: /org/bluealsa
bluealsa: [178] D: bluez.c:153: Registering media application: /org/bluez/hci0
bluealsa: [178] D: bluez.c:743: Exporting media endpoint object: /org/bluez/hci0/A2DP/AAC/sink/1
bluealsa: [178] D: bluez.c:743: Exporting media endpoint object: /org/bluez/hci0/A2DP/AAC/sink/2
bluealsa: [178] D: bluez.c:743: Exporting media endpoint object: /org/bluez/hci0/A2DP/SBC/sink/1
bluealsa: [178] D: bluez.c:743: Exporting media endpoint object: /org/bluez/hci0/A2DP/SBC/sink/2
bluealsa: [178] D: bluez.c:1290: Signal: org.freedesktop.DBus.ObjectManager.InterfacesAdded(/org/bluez/hci0/dev_60_00_01_02_03_04, ...)
bluealsa: [178] D: bluez.c:1290: Signal: org.freedesktop.DBus.ObjectManager.InterfacesAdded(/org/bluez/hci0/dev_60_00_01_02_03_04/fd0, ...)
bluealsa: [178] D: dbus.c:47: Called: org.bluez.MediaEndpoint1.SetConfiguration() on /org/bluez/hci0/A2DP/AAC/sink/1
bluealsa: [178] D: storage.c:123: Loading storage: //lib/bluealsa/60:00:01:02:03:04
bluealsa: [178] D: bluez.c:572: A2DP Sink (AAC) configured for device 60:00:01:02:03:04
bluealsa: [178] D: bluez.c:575: A2DP selected configuration blob [len=6]: 800104835b60
bluealsa: [178] D: bluez.c:577: PCM configuration: channels: 2, sampling: 44100
bluealsa: [178] D: bluez.c:743: Exporting media endpoint object: /org/bluez/hci0/A2DP/AAC/sink/3
bluealsa: [178] D: bluez.c:1290: Signal: org.freedesktop.DBus.ObjectManager.InterfacesAdded(/org/bluez/hci0/dev_60_00_01_02_03_04/sep2, ...)
bluealsa: [178] D: bluez.c:1362: Adding new Stream End-Point: 60:00:01:02:03:04: SRC: AAC

bluealsa: [212] D: dbus.c:47: Called: org.bluealsa.PCM1.Open() on /org/bluealsa/hci0/dev_60_00_01_02_03_04/a2dpsnk/source
bluealsa: [178] D: ba-transport-pcm.c:453: Closing PCM: 15
bluealsa: [202] D: ba-transport.c:203: PCM clients check keep-alive: 0 ms
bluealsa: [178] E: main.c:141: Lost BlueALSA D-Bus name: org.bluealsa
bluealsa: [178] D: ba-transport.c:826: Freeing transport: A2DP Sink (AAC)
bluealsa: [178] D: storage.c:160: Saving storage: //lib/bluealsa/60:00:01:02:03:04
bluealsa: [178] D: ba-device.c:143: Freeing device: 60:00:01:02:03:04
bluealsa: [178] D: ba-adapter.c:144: Freeing adapter: hci0
```
```
# bluealsa-aplay -v -Dplughw:CARD=Headphones --profile-a2dp
bluealsa-aplay: [210] I: aplay.c:1270: Selected configuration:
  BlueALSA service: org.bluealsa
  ALSA PCM device: plughw:CARD=Headphones
  ALSA PCM buffer time: 500000 us
  ALSA PCM period time: 100000 us
  Volume control type: auto
  ALSA mixer device: default
  ALSA mixer element: 'Master',0
  Bluetooth device(s): ANY
  Profile: A2DP
bluealsa-aplay: [210] D: aplay.c:904: Creating IO worker 60:00:01:02:03:04
bluealsa-aplay: [210] D: aplay.c:1320: Starting main loop
bluealsa-aplay: [211] D: aplay.c:577: Opening BlueALSA source PCM: /org/bluealsa/hci0/dev_60_00_01_02_03_04/a2dpsnk/source
bluealsa-aplay: [211] E: aplay.c:580: Couldn't open BlueALSA source PCM: Message recipient disconnected from message bus without replying
bluealsa-aplay: [211] D: aplay.c:536: Exiting IO worker 60:00:01:02:03:04
```

```
[bluetooth]# show                                                                  
Controller B8:27:EB:00:01:02 (public)
	Manufacturer: 0x000f (15)
	Version: 0x07 (7)
	Name: lithium
	Alias: lithium
	Class: 0x00040414 (263188)
	Powered: yes
	PowerState: on
	Discoverable: yes
	DiscoverableTimeout: 0x00000000 (0)
	Pairable: yes
	UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
	UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
	UUID: PnP Information           (00001200-0000-1000-8000-00805f9b34fb)
	UUID: Audio Sink                (0000110b-0000-1000-8000-00805f9b34fb)
	UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
	Modalias: usb:v1D6Bp0246d054F
	Discovering: no
	Roles: central
	Roles: peripheral
Advertising Features:
	ActiveInstances: 0x00 (0)
	SupportedInstances: 0x05 (5)
	SupportedIncludes: tx-power
	SupportedIncludes: appearance
	SupportedIncludes: local-name
	SupportedCapabilities.MaxAdvLen: 0x1f (31)
	SupportedCapabilities.MaxScnRspLen: 0x1f (31)
[bluetooth]# hci0 60:00:01:02:03:04 type BR/EDR connected eir_len 15               
[bluetooth]# [NEW] Device 60:00:01:02:03:04 Phone
[Phone]# Request confirmation
[agent] Confirm passkey 579388 (yes/no): yes
[Phone]# hci0 new_link_key 60:00:01:02:03:04 type 0x08 pin_len 0 store_hint 1   
[Phone]# hci0 device_flags_changed: 60:00:01:02:03:04 (BR/EDR)
[Phone]#      supp: 0x00000000  curr: 0x00000000
[Phone]# [CHG] Device 60:00:01:02:03:04 Bonded: yes
[Phone]# [CHG] Device 60:00:01:02:03:04 AddressType: public
[Phone]# [CHG] Device 60:00:01:02:03:04 Modalias: bluetooth:v0075p0100d0201
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001105-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000110a-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000110c-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000110e-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001112-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001115-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001116-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000111f-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000112d-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000112f-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001132-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001200-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001800-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001801-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 ServicesResolved: yes
[Phone]# [CHG] Device 60:00:01:02:03:04 Paired: yes
[Phone]# Authorize service
yes
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001105-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000110a-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000110c-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000110d-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000110e-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001112-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001115-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001116-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000111f-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000112d-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 0000112f-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001132-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001200-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001800-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Device 60:00:01:02:03:04 UUIDs: 00001801-0000-1000-8000-00805f9b34fb
[Phone]# [NEW] Transport /org/bluez/hci0/dev_60_00_01_02_03_04/fd0 
[Phone]# [NEW] Endpoint /org/bluez/hci0/dev_60_00_01_02_03_04/sep2 
[Phone]# [DEL] Transport /org/bluez/hci0/dev_60_00_01_02_03_04/fd0 
[Phone]# hci0 class of device changed: 0x000414
[Phone]# [CHG] Controller B8:27:EB:00:01:02 Class: 0x00000414 (1044)
[Phone]# [CHG] Controller B8:27:EB:00:01:02 UUIDs: 00001801-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Controller B8:27:EB:00:01:02 UUIDs: 00001800-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Controller B8:27:EB:00:01:02 UUIDs: 00001200-0000-1000-8000-00805f9b34fb
[Phone]# [CHG] Controller B8:27:EB:00:01:02 UUIDs: 0000180a-0000-1000-8000-00805f9b34fb
[Phone]# [DEL] Endpoint /org/bluez/hci0/dev_60_00_01_02_03_04/sep2 
[Phone]# hci0 60:00:01:02:03:04 type BR/EDR disconnected with reason 3
[Phone]# [CHG] Device 60:00:01:02:03:04 ServicesResolved: no
[bluetooth]# [CHG] Device 60:00:01:02:03:04 Connected: no
```

kernel miniconfig (allnoconfig being the basis)
```
CONFIG_DEFAULT_HOSTNAME="test"
CONFIG_DEFAULT_MMAP_MIN_ADDR=32768
CONFIG_SMP=y
CONFIG_CPU_FREQ=y
CONFIG_NR_CPUS=4
CONFIG_ARCH_BCM=y
CONFIG_ARCH_BCM2835=y
CONFIG_DMA_CMA=y
CONFIG_DMADEVICES=y
CONFIG_DMA_BCM2835=y
CONFIG_MAILBOX=y
CONFIG_BCM2835_MBOX=y
CONFIG_RASPBERRYPI_FIRMWARE=y
CONFIG_CLK_BCM2835=y
CONFIG_CLK_RASPBERRYPI=y
CONFIG_WATCHDOG=y
CONFIG_BCM2835_WDT=y
CONFIG_HW_RANDOM=y
CONFIG_HW_RANDOM_BCM2835=y
CONFIG_PM=y
CONFIG_BCM2835_POWER=y
CONFIG_RASPBERRYPI_POWER=y
CONFIG_THERMAL=y
CONFIG_THERMAL_OF=y
CONFIG_BCM2835_THERMAL=y
CONFIG_USB_DEFAULT_PERSIST=y
CONFIG_USB_DWC2=y
CONFIG_GPIO_RASPBERRYPI_EXP=y
CONFIG_SERIAL_8250=y
CONFIG_SERIAL_8250_CONSOLE=y
CONFIG_SERIAL_8250_EXTENDED=y
CONFIG_SERIAL_8250_SHARE_IRQ=y
CONFIG_SERIAL_8250_BCM2835AUX=y
CONFIG_SERIAL_OF_PLATFORM=y
CONFIG_NETDEVICES=y
CONFIG_USB_SUPPORT=y
CONFIG_USB=y
CONFIG_USB_NET_DRIVERS=y
CONFIG_USB_USBNET=y
CONFIG_USB_NET_SMSC95XX=y
CONFIG_IP_PNP=y
CONFIG_IP_PNP_DHCP=y
CONFIG_DEVTMPFS=y
CONFIG_SYSFS=y
CONFIG_CGROUPS=y
CONFIG_LOG_BUF_SHIFT=19
CONFIG_RELOCATABLE=y
CONFIG_CRYPTO_SHA256=y
CONFIG_KEXEC_FILE=y
CONFIG_EXPERT=y
CONFIG_MULTIUSER=y
CONFIG_FHANDLE=y
CONFIG_POSIX_TIMERS=y
CONFIG_PRINTK=y
CONFIG_FUTEX=y
CONFIG_EPOLL=y
CONFIG_SIGNALFD=y
CONFIG_TIMERFD=y
CONFIG_EVENTFD=y
CONFIG_ADVISE_SYSCALLS=y
CONFIG_BINFMT_ELF=y
CONFIG_BINFMT_SCRIPT=y
CONFIG_NET=y
CONFIG_UNIX=y
CONFIG_INET=y
CONFIG_PROC_FS=y
CONFIG_PROC_SYSCTL=y
CONFIG_INOTIFY_USER=y
CONFIG_FILE_LOCKING=y
CONFIG_FANOTIFY=y
CONFIG_NLS=y
CONFIG_NLS_DEFAULT="utf8"
CONFIG_NLS_UTF8=y
CONFIG_STRIP_ASM_SYMS=y
CONFIG_JUMP_LABEL=y
CONFIG_HIGH_RES_TIMERS=y
CONFIG_CMA=y
CONFIG_PANIC_ON_OOPS=y
CONFIG_PANIC_TIMEOUT=10
CONFIG_CONSOLE_LOGLEVEL_QUIET=2
CONFIG_RTC_HCTOSYS=y
CONFIG_RTC_CLASS=y
CONFIG_RTC_SYSTOHC=y
CONFIG_UNIX98_PTYS=y
CONFIG_CONSOLE_LOGLEVEL_DEFAULT=8
CONFIG_RFKILL=y
CONFIG_RFKILL_GPIO=y
CONFIG_SOUND=y
CONFIG_SND=y
CONFIG_SND_PCM_TIMER=y
CONFIG_SND_CTL_FAST_LOOKUP=y
CONFIG_SND_DRIVERS=y
CONFIG_SND_ALOOP=y
CONFIG_SND_HDA_HWDEP=y
CONFIG_SND_HDA_RECONFIG=y
CONFIG_SND_SOC=y
CONFIG_SND_SOC_BT_SCO=y
CONFIG_SND_SIMPLE_CARD=y
CONFIG_SND_AUDIO_GRAPH_CARD=y
CONFIG_SND_AUDIO_GRAPH_CARD2=y
CONFIG_SND_AUDIO_GRAPH_CARD2_CUSTOM_SAMPLE=y
CONFIG_SND_TEST_COMPONENT=y
CONFIG_SND_SOC_SIMPLE_AMPLIFIER=y
CONFIG_SND_SOC_SIMPLE_MUX=y
CONFIG_SND_SOC_HDA=y
CONFIG_SND_SOC_AC97_CODEC=y
CONFIG_SND_SOC_ADAU7002=y
CONFIG_SND_SOC_ADAU7118_HW=y
CONFIG_SND_PROC_FS=y
CONFIG_SND_VERBOSE_PROCFS=y
CONFIG_STAGING=y
CONFIG_BCM_VIDEOCORE=y
CONFIG_VCHIQ_CDEV=y
CONFIG_BCM2835_VCHIQ_MMAL=y
CONFIG_SND_BCM2835=y
CONFIG_BT_BREDR=y
CONFIG_BT_LE=y
CONFIG_INPUT=y
CONFIG_INPUT_MISC=y
CONFIG_INPUT_UINPUT=y
CONFIG_CRYPTO=y
CONFIG_CRYPTO_MANAGER=y
CONFIG_CRYPTO_USER=y
CONFIG_CRYPTO_PCRYPT=y
CONFIG_CRYPTO_CRYPTD=y
CONFIG_CRYPTO_AES=y
CONFIG_CRYPTO_AES_TI=y
CONFIG_CRYPTO_KEYWRAP=y
CONFIG_CRYPTO_CMAC=y
CONFIG_CRYPTO_AEGIS128=y
CONFIG_CRYPTO_AEGIS128_SIMD=y
CONFIG_CRYPTO_CCM=y
CONFIG_CRYPTO_GCM=y
CONFIG_CRYPTO_SEQIV=y
CONFIG_CRYPTO_ECHAINIV=y
CONFIG_CRYPTO_ESSIV=y
CONFIG_CRYPTO_USER_API_HASH=y
CONFIG_CRYPTO_USER_API_SKCIPHER=y
CONFIG_CRYPTO_USER_API_AEAD=y
CONFIG_CRYPTO_AES_ARM64_NEON_BLK=y
CONFIG_CRYPTO_AES_ARM64_BS=y
CONFIG_CRYPTO_AES_ARM64_CE=y
CONFIG_CRYPTO_AES_ARM64_CE_BLK=y
CONFIG_CRYPTO_AES_ARM64=y
CONFIG_CRYPTO_AES_ARM64_CE_CCM=y
CONFIG_SYSVIPC=y
CONFIG_SERIAL_DEV_BUS=y
CONFIG_BT=y
CONFIG_BT_HCIUART=y
CONFIG_BT_HCIUART_BCM=y
CONFIG_TTY=y
CONFIG_SERIAL_AMBA_PL011=y
CONFIG_SERIAL_AMBA_PL011_CONSOLE=y
CONFIG_LOCALVERSION="-test"
CONFIG_BLK_DEV_INITRD=y
CONFIG_INITRAMFS_SOURCE="root.list.sort"
CONFIG_CMDLINE="console=ttyS1,115200 ip=dhcp"
CONFIG_CMDLINE_FORCE=y
```

**Comment by borine** [comment]
> bluetoothd -p a2dp

Why are you using that option? Try without the `-p a2dp` then you may have more luck.

**Comment by mwetterw** [comment]
> > bluetoothd -p a2dp
> 
> Why are you using that option? Try without the `-p a2dp` then you may have more luck. 

At first I did many attempts without restricting the Bluetooth profiles loaded by bluetoothd.
But this makes the situation less clear and creates problems:
`bluetoothd` said "I support many things" to the Bluetooth world. For example, it advertised AVRCP, so my Phone tried to talk AVRCP with bluetoothd.
It triggered issues such has "/dev/input" missing.
That's why I ended up activating `CONFIG_INPUT_UINPUT` in my kernel config to solve this issue.

So to sum up, I restricted the profiles to reduce the scope of the problem in order to see clearer and tackle one problem at a time.
For now, the only goal is to hear the music being played, not be able to use remote control.
Once that issue is solved, I will turn on more profiles, more codecs, etc.

**Comment by borine** [comment]
The problem line, for me, in your logs is the bluealsa log entry:
```
bluealsa: [178] E: main.c:141: Lost BlueALSA D-Bus name: org.bluealsa
```
which is occurring when bluealsa tries to open the AVDTP transport stream endpoint. That is odd, why would D-Bus unregister the service name at that point?
I've never seen that before, so my guess is that some other bluetooth function is at play. Try adding other likely bluetooth plugins, one at a time, to see if any make a difference. I would start with the policy plugin:
```
bluetoothd -n -p a2dp,policy
```
Beyond that, I don't see any other clue in the logs as to what is happening here.

**Comment by arkq** [comment]
> The problem line, for me, in your logs is the bluealsa log entry:
> `bluealsa: [178] E: main.c:141: Lost BlueALSA D-Bus name: org.bluealsa`

Yes, this seems to be the issue here (at least the very last stage of the issue). To investigate it better please collect logs from `dbus-monitor --system` as root. The important part should be just before the name is lost (maybe few messages before). I've seen things like that and mostly they were related to some issues with D-Bus message processing within a code which uses glib for handling D-Bus connection (message was corrupted which cases connection to be dropped). However, it might be something else like some "required" plugin not being loaded by BlueZ.

**Comment by mwetterw** [comment]
Thanks for your help!

You'll find the whole dbus conversation [here](https://github.com/user-attachments/files/18041622/dbus.txt).
I've analyzed the whole dbus session.
My own decoding of the list of "dbus actors" in play:
```
1.0:            bluez
1.5, 1.7:       bluetoothctl
1.6:            bluealsa
1.8, 1.9, 1.10: bluealsa-aplay
```

The whole dbus log seems logical. Except here:

> method call time=1733440786.944519 sender=:1.10 -> destination=org.bluealsa serial=5 path=/org/bluealsa; interface=org.freedesktop.DBus.ObjectManager; member=GetManagedObjects
> method return time=1733440786.946730 sender=:1.6 -> destination=:1.10 serial=22 reply_serial=5
>    array [
>       dict entry(
>          object path "/org/bluealsa/hci0/dev_60_00_01_02_03_04/a2dpsnk/source"
>          array [
>             dict entry(
>                string "org.bluealsa.PCM1"
>                array [
>                   dict entry(
>                      string "Device"
>                      variant                         object path "/org/bluez/hci0/dev_60_00_01_02_03_04"
>                   )
>                   dict entry(
>                      string "Sequence"
>                      variant                         uint32 0
>                   )
>                   dict entry(
>                      string "Transport"
>                      variant                         string "A2DP-sink"
>                   )
> [...]
> **method call time=1733440786.949010 sender=:1.10 -> destination=org.bluealsa serial=6 path=/org/bluealsa/hci0/dev_60_00_01_02_03_04/a2dpsnk/source; interface=org.bluealsa.PCM1; member=Open**
> signal time=1733440786.956571 sender=org.freedesktop.DBus -> destination=:1.6 serial=11 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=**NameLost**
>    string "org.bluealsa"
> signal time=1733440786.956693 sender=org.freedesktop.DBus -> destination=(null destination) serial=14 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=**NameOwnerChanged**
>    string "org.bluealsa"
>    string ":1.6"
>    string ""
> signal time=1733440786.959041 sender=org.freedesktop.DBus -> destination=:1.6 serial=12 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=**NameLost**
>    string ":1.6"
> signal time=1733440786.959138 sender=org.freedesktop.DBus -> destination=(null destination) serial=20 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=**NameOwnerChanged**
>    string ":1.6"
>    string ":1.6"
>    string ""
> **error time=1733440786.959229 sender=org.freedesktop.DBus -> destination=:1.10 error_name=org.freedesktop.DBus.Error.NoReply reply_serial=6
>    string "Message recipient disconnected from message bus without replying"**

It appears that `bluealsa` crashes just after being asked by `bluealsa-aplay` to `Open` the `org.bluealsa.PCM1` `org/bluealsa/hci0/dev_60_00_01_02_03_04/a2dpsnk/source`.

It is not yet clear for me if the loss of the DBus name is a cause or a consequence.
Isn't it just a consequence? Maybe the name is lost because `bluealsa` crashed when trying to do the actions of opening the PCM.
I think we should zoom in in the `bluealsa` code portion opening the PCM and add debug messages.

**Comment by arkq** [comment]
Ok, so now, if you can, try to run bluealsa under gdb. You can add breakpoint in the function which logs name lost and when stopped print back trace. Maybe this will show something. If this will not show anything useful then it might be required to add some additional logs or breakpoints to help track the root cause...

**Comment by borine** [comment]
> Maybe the name is lost because bluealsa crashed when trying to do the actions of opening the PCM

If bluealsa crashed *before* the name lost event, then there cannot be a "Name Lost" error message in the bluealsa log - but there is such a message in the bluealsa log.

also at the end of the bluealsa log:
```
bluealsa: [178] D: ba-adapter.c:144: Freeing adapter: hci0
```
indicates that the bluealsa daemon does not crash, but shuts down cleanly.

**Comment by mwetterw** [comment]
> Ok, so now, if you can, try to run bluealsa under gdb. You can add breakpoint in the function which logs name lost and when stopped print back trace. Maybe this will show something. If this will not show anything useful then it might be required to add some additional logs or breakpoints to help track the root cause...
> 
> EDIT: Before running `bluealsa` with `gdb` please install debug symbols for `libc` and `glib`. Otherwise we will see nothing useful.

I know a bit more. I found the `if` in `bluealsa`'s code making it shutdown.

[bluealsa_pcm_open()](https://github.com/arkq/bluez-alsa/blob/v4.3.1/src/bluealsa-dbus.c#L426) is the handler that is called when `bluealsa-aplay` asks `bluez` via D-Bus to open the PCM.
It fails when calling [ba_transport_pcm_signal_send()](https://github.com/arkq/bluez-alsa/blob/v4.3.1/src/bluealsa-dbus.c#L514).
`ba_transport_pcm_signal_send()` fails because of [this if](https://github.com/arkq/bluez-alsa/blob/v4.3.1/src/ba-transport-pcm.c#L546) and so the function returns `-1`.
So if I understand correctly, we are trying to send a signal to a transport PCM whose state is not running.

Note that I try to launch `bluealsa-aplay` once the Phone is connected, but before the Phone plays any sound.
Is it correct? Or should I first play sound with the Phone, and then start `bluealsa-aplay`?
If I first play sound with the phone before launching `bluealsa-aplay`, I get another error:

```
bluealsa: [251] D: bluez.c:1480: Signal: org.freedesktop.DBus.Properties.PropertiesChanged(): org.bluez.MediaTransport1: State
bluealsa: [251] E: ba-transport.c:327: Couldn't acquire transport: GDBus.Error:org.bluez.Error.Failed: I/O error
bluealsa: [251] D: bluez.c:1480: Signal: org.freedesktop.DBus.Properties.PropertiesChanged(): org.bluez.MediaTransport1: State
bluealsa: [251] D: bluez.c:1480: Signal: org.freedesktop.DBus.Properties.PropertiesChanged(): org.bluez.MediaTransport1: State
bluealsa: [251] E: ba-transport.c:327: Couldn't acquire transport: GDBus.Error:org.bluez.Error.NotAuthorized: Operation Not Authorized
```

I hope this will help us getting closer and closer to my issue. :)

**Comment by borine** [comment]
> should I first play sound with the phone, and then start bluealsa-aplay?

you can start bluealsa-aplay whenever you want. I always start it at boot and just leave it running until shutdown.


> It fails when calling ba_transport_pcm_signal_send().

No, that is not a failure. An A2DP sink cannot start an audio stream, it has to wait for the A2DP source device (ie the phone) to do that. The point of sending that signal is that the PCM I/O thread may already be running (e.g. because of the keep-alive option after the previous client connection); but in this case there is no existing audio stream from the phone, so there can be no I/O thread running and so nowhere to deliver that signal.

> bluealsa: [251] E: ba-transport.c:327: Couldn't acquire transport: GDBus.Error:org.bluez.Error.NotAuthorized: Operation Not Authorized

So now we have a genuine error. Bluez is not permitting bluealsa to acquire the transport. That sounds very much like an error in the system config. The Bluez documentation is not very helpful:
https://github.com/bluez/bluez/blob/master/doc/org.bluez.MediaTransport.rst#fd-uint16-uint16-acquire - it just states that it can fail with org.bluez.Error.NotAuthorized without giving any reason for possible causes. So I can't give any hints on where to start looking, but I am still convinced that this issue is related to policy management either in Bluez or DBus, or both.

**Comment by mwetterw** [comment]
I finally found the root-cause of my issue!
It was very hard to find. I'm actually lucky to find this now.

I was missing `#define HAVE_UNIX_FD_PASSING 1` in `config.h` when compiling the **D-Bus**.

Thank you again for your help.
